lib/services/ngsi/entities-NGSI-v2.js

```diff
@@ -489,176 +375,580 @@ function sendUpdateValueNgsi2(entityName, attributes, typeInformation, token, ca
+                    return;
+                }
+            }
+        }
+    }
+    logger.debug(context, 'sendUpdateValueNgsi2Refactored \n ending payload=%j', payload);
+
+    for (let m = 0; m < payload.entities.length; m++) {
+        for (var key in payload.entities[m]) {
+            // purge object_id from payload
+            if (payload.entities[m][key] && payload.entities[m][key].object_id) {
+                delete payload.entities[m][key].object_id;
             }
         }
+        payload.entities[m] = NGSIUtils.castJsonNativeAttributes(payload.entities[m]); // native types
+    }
+    logger.debug(
+        context,
+        'sendUpdateValueNgsi2Refactored \n payload with native types and without object_id %j',
+        payload
     );
+
+    options.json = payload;
+
+    // Prevent to update an entity with an empty payload
+    if (
+        Object.keys(options.json).length > 0 &&
+        (options.json.entities.length > 1 ||
+            (options.json.entities.length === 1 && Object.keys(options.json.entities[0]).length > 2)) // more than id and type
+    ) {
+        // Final check: (to keep tests unchanged) before do CB requests
+        //              one entity     -> request /v2/entities/ + entityName + /atts ?type=typeInformation.type
+        //              multi entities -> request /v2/op/update
+        if (options.json.entities.length === 1) {
+            // recreate options object to use single entity update
+            url = '/v2/entities/' + entityName + '/attrs';
+            if (typeInformation && typeInformation.type) {
+                url += '?type=' + typeInformation.type;
+            }
+            options = NGSIUtils.createRequestObject(url, typeInformation, token);
+            options.json = payload.entities[0];
+            delete options.json.id;
+            delete options.json.type;
+            if (config.getConfig().appendMode === true) {
+                options.method = 'POST';
+            } else {
+                options.method = 'PATCH';
+            }
+        } // else: keep current options object created for a batch update
+        logger.debug(context, 'Updating device value in the Context Broker at [%s]', options.url);
+        logger.debug(context, 'Using the following NGSI v2 request:\n\n%s\n\n', JSON.stringify(options, null, 4));
+        request(
+            options,
+            generateNGSI2OperationHandler('update', entityName, typeInformation, token, options, callback)
+        );
+    } else {
+        logger.debug(
+            context,
+            'Not updating device value in the Context Broker at [%s] due to empty payload \n\n[%s]\n\n',
+            options.url,
+            JSON.stringify(options, null, 4)
+        );
+        callback(null);
+    }
 }
 
 exports.sendQueryValue = sendQueryValueNgsi2;
-exports.sendUpdateValue = sendUpdateValueNgsi2;
+exports.sendUpdateValue = sendUpdateValueNgsi2Refactored;
```

Comment:
leftover?

Reply:
It would be a good idea to a case insensitive search for "refactor" string in the code base to ensure none of theses is leaked to the final version of the code.